app/utils/lava_log_parser.py

```diff
@@ -0,0 +1,126 @@
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Based on log2html.py:
+# https://github.com/khilman/lab/blob/master/log2html.py
+#
+# Error/warning formatting, css, colors etc. stolen from
+# Olof Johansson <olof@lixom.net>
+
+import argparse
+import cgi
+import json
+import yaml
+try:
+    import cStringIO as StringIO
+except ImportError:
+    import StringIO
+
+HTML_HEAD = """\
+<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
+<html>
+<head>
+  <title>Boot log: {board}</title>
+  <style type="text/css">
+  body {{ background-color: black; color: white; }}
+  pre {{ font-size: 0.8em; }}
+  span.pass {{ color: green; }}
+  span.err {{ color: red; }}
+  span.warn {{ color: #F88017; }}
+  a:link {{text-decoration: none }}
+  a:visited {{text-decoration: none }}
+  a:active {{text-decoration: none }}
+  a:hover {{text-decoration: bold; color: red; }}
+  ul.results {{
+    list-style-type: none;
+    padding-left: 0;
+    font-weight: bold;
+    font-size: 1.3em;
+    font-family: "Courier New", Courier, monospace;
+  }}
+  li.result {{ padding: 5px 0 5px 0; }}
+  </style>
+</head>
+<body>
+<h1>Boot log: {board}</h1>
+"""
+
+
+def run(log, boot, txt, html):
+    boot_result = boot.get('boot_result', 'Unknown')
+    if boot_result == "PASS":
+        boot_result_html = "<span class=\"pass\">PASS</span>"
+    elif boot_result == "FAIL":
+        boot_result_html = "<span class=\"err\">FAIL</span>"
+    else:
+        boot_result_html = "<span class=\"warn\">{}</span>".format(boot_result)
+
+    formats = {
+        "warning": "<span class=\"warn\">{}</span>\n",
+        "error": "<span class=\"err\">{}</span>\n",
+    }
+    numbers = {'warning': 0, 'error': 0}
+    log_buffer = StringIO.StringIO()
+
+    for line in log:
+        level = line['lvl']
+        fmt = formats.get(level, None)
+        if fmt:
+            log_buffer.write(fmt.format(line['msg']))
+            numbers[level] += 1
+        elif level == 'target':
+            msg = line['msg']
+            log_buffer.write(msg)
+            log_buffer.write('\n')
+            txt.write(msg)
+            txt.write('\n')
+
+    html.write(HTML_HEAD.format(board=boot['board']))
+    html.write("<ul class=\"results\">")
+    results = {
+        "Boot result": boot_result_html,
+        "Errors": numbers['error'],
+        "Warnings": numbers['warning'],
+    }
+    for title, value in results.iteritems():
+        html.write("<li class=\"result\">{}: {}</li>".format(title, value))
+    html.write("</ul><pre>\n")
+    html.write(cgi.escape(log_buffer.getvalue()))
```

Comment:
@MiloCasagrande Calling cgi.escape solves a lot of issues with special HTML characters:
https://people.collabora.com/~gtucker/tmp/kernelci-fail-escaped.html
Before:
https://people.collabora.com/~gtucker/tmp/kernelci-fail.html